### PR TITLE
FPGrowth/FPMax and Association Rules with the existence of missing values (#1004)

### PR DESCRIPTION
**Pull request by zazass8**
### Description

This update intends to implement the FP-Growth and FP-Max algorithms from frequent_patterns with the possibility of missing values in the input dataset. The code implements the same structure and logic of the algorithms, while computing the support metric as in "ignoring" the missing values in the data. That gives a more realistic indication of the frequency of existence in the items/itemsets that are generated from the algorithm. Given the output of the algorithm, the corresponding association rules and metrics are also updated taking into account the existence of missing values.

The input accepted for this implementation is a `pandas.DataFrame` that accepts binary input as 0/1 or True/False (as it was originally), as well as 0/1/NaN or True/False/NaN, where NaN is `numpy.nan`.

Please also find attached the corresponding paper conducting this research of association rule mining with the existence of missing values.
[ragel1998.pdf](https://github.com/user-attachments/files/17276897/ragel1998.pdf)

### Related issues or pull requests

Please use [this link](https://github.com/rasbt/mlxtend/issues/1004) for the corresponding discussion of this new feature in the issue tracker.

### Pull Request Checklist

- [x] Added a note about the modification or contribution to the `./docs/sources/CHANGELOG.md` file (if applicable)
- [x] Added appropriate unit test functions in the `./mlxtend/*/tests` directories (if applicable)
- [x] Modify documentation in the corresponding Jupyter Notebook under `mlxtend/docs/sources/` (if applicable)
- [x] Ran `PYTHONPATH='.' pytest ./mlxtend -sv` and make sure that all unit tests pass (for small modifications, it might be sufficient to only run the specific test file, e.g., `PYTHONPATH='.' pytest ./mlxtend/classifier/tests/test_stacking_cv_classifier.py -sv`)
- [x] Checked for style issues by running `flake8 ./mlxtend`

### CODE REVIEW

**Comment by zazass8** [comment]
@rasbt there's an issue with CI, as you can see in fpcommon.py line 34 the function np.nansum() is not passing several tests whereas testing it with np.sum() passes all CI tests but practically it gives the wrong answers. I think it may require to make some changes to the tests. I tried already to change the code in fpommon.py but it seems np.nansum() is the one that is suitable for our purpose. By counting the 1s while "ignoring" NaNs. Let me know if you have in mind an alternative solution.

**Comment by zazass8** [comment]
@rasbt In metrics such as Zhang's and Jaccard, if `null_values=True` it computes values that are outside the unrestricted range. Sometimes it occurs with the certainty metric as well. I haven't found anywhere in the literature if there's a modified formula of these metrics that also takes into account the existence of NaNs. If there is, I am happy to update otherwise if you prefer I could exclude these metrics from the final array that is created. Let me know :)

File: docs/sources/CHANGELOG.md [hunk 1]
```diff
@@ -18,11 +18,18 @@ The CHANGELOG for the current development version is available at
...
+  - [`mlxtend.frequent_patterns.association_rules`](https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/)
+- [`mlxtend.frequent_patterns.association_rules`](https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/)Implemented three new metrics: Jaccard, Certainty, and Kulczynski. ([#1096](https://github.com/rasbt/mlxtend/issues/1096))
+- Integrated scikit-learn's `set_output` method into `TransactionEncoder` ([#1087](https://github.com/rasbt/mlxtend/issues/1087) via[it176131](https://github.com/it176131))
 
 ##### Changes
 
+- [`mlxtend.frequent_patterns.fpcommon`] Added the null_values parameter in valid_input_check signature to check in case the input also includes null values. Changes the returns statements and function signatures for setup_fptree and generated_itemsets respectively to return the disabled array created and to include it as a parameter. Added code in [`mlxtend.frequent_patterns.fpcommon`] and [`mlxtend.frequent_patterns.association_rules`](https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/) to implement the algorithms in case null values exist when null_values is True.
```
Comment (rasbt) [inline docs/sources/CHANGELOG.md:32]:
```suggestion
- [`mlxtend.frequent_patterns.fpcommon`] Added the `null_values` parameter in `valid_input_check` signature to check in case the input also includes null values. Changes the returns statements and function signatures for `setup_fptree` and `generated_itemsets` respectively to return the disabled array created and to include it as a parameter. Added code in [`mlxtend.frequent_patterns.fpcommon`] and [`mlxtend.frequent_patterns.association_rules`](https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/) to implement the algorithms in case null values exist when `null_values` is `True`.
```

**Comment by rasbt** [comment]
Just wanted to say thanks for the PR! I first bookmarked it 2 weeks ago, but then I got sick and am just slowly catching up.

File: mlxtend/frequent_patterns/fpgrowth.py [hunk 2]
```diff
@@ -42,6 +44,9 @@ def fpgrowth(df, min_support=0.5, use_colnames=False, max_len=None, verbose=0):
       The support is computed as the fraction
       transactions_where_item(s)_occur / total_transactions.
 
+    null_values : bool (default: True)
```
Comment (rasbt) [inline mlxtend/frequent_patterns/fpgrowth.py:47]:
```suggestion
    null_values : bool (default: False)
```

I think this should say `False` based on the function signature above. (This is a great feature, but `False` is probably also the right choice for backward compatibility.)

File: mlxtend/frequent_patterns/association_rules.py [hunk 3]
```diff
@@ -30,6 +33,9 @@
 
 def association_rules(
     df: pd.DataFrame,
+    df_or: pd.DataFrame,
```
Comment (rasbt) [inline mlxtend/frequent_patterns/association_rules.py:36]: If I see it correctly, the DataFrame is only required if `null_values` is `True`, correct? In this case, we could also make this argument optional and set it to `None` by default for backwards compatibility. (And then if someone saids `null_values=True` we raise an error that `df_or` should be provided.

Perhaps we should also rename `df_or` to `df_orig`. The name `df_or` made be first think of a logical or like `df_and` and `df_or`. What do you think?

**Comment by zazass8** [comment]
No worries at all, my pleasure to contribute on your project! No worries at all, I am glad you are feeling better now. I resolved and made the changes you mentioned above, feel free to have another look and let me know if there's anything else!

File: mlxtend/frequent_patterns/fpcommon.py [hunk 4]
```diff
@@ -163,6 +164,19 @@ def valid_input_check(df, null_values=False):
             "Please use a DataFrame with bool type",
             DeprecationWarning,
         )
+
+        # If null_values is True but no NaNs are found, raise an error
```
Comment (rasbt) [inline mlxtend/frequent_patterns/fpcommon.py:168]: Just curious, would this lead to inaccurate metrics or is it just inefficient. Or are there other reasons why this would not be permitted?

File: mlxtend/frequent_patterns/tests/test_association_rules.py [hunk 5]
```diff
@@ -55,7 +55,7 @@
 
 # fmt: off
 def test_default():
-    res_df = association_rules(df_freq_items, df, len(df))
+    res_df = association_rules(df_freq_items, len(df), df)
```
Comment (rasbt) [inline mlxtend/frequent_patterns/tests/test_association_rules.py:58]: If NaN values are not present, the unit tests don't need the extra `df` here though, correct? I wonder if we could have the unit tests that don't have NaN values as they were before but then have an additional one where `df` is provided as you showed just to check that it doesn't cause errors. What do you think?

**Comment by rasbt** [comment]
> No worries at all, my pleasure to contribute on your project! No worries at all, I am glad you are feeling better now. I resolved and made the changes you mentioned above, feel free to have another look and let me know if there's anything else!

Thanks for addressing these! I just had 2 follow-up comments above, but no rush.

File: mlxtend/frequent_patterns/fpcommon.py (hunk 4 above)
Comment (zazass8) [reply mlxtend/frequent_patterns/fpcommon.py:168]: No, this one is just inefficient. Normally if I don't have an array with null values but null_values is True, it'd generate the final dataframe with metrics when null values do not exist (as original approach). The only reason I included is to remain consistent and still raise the error if it is the case. But since null_values is set False as default, maybe there's no point of including this condition as the user may not intend to set null_values to True if there are no NaNs in their array. Up to you, if you want I can completely exclude it.

File: mlxtend/frequent_patterns/tests/test_association_rules.py (hunk 5 above)
Comment (zazass8) [reply mlxtend/frequent_patterns/tests/test_association_rules.py:58]: We can definitely do that, it is a great idea! I removed `df` from the signature but still kept `num_itemsets=len(df)` as it is necessary even for the test case when NaNs do not exist. I also added a `test_nullability()` method to test the case when null values exist. Let me know if you need anything else.

File: mlxtend/frequent_patterns/fpcommon.py (hunk 4 above)
Comment (rasbt) [reply mlxtend/frequent_patterns/fpcommon.py:168]: Thanks, that's what I thought! Sounds good! In this case, maybe we should not raise an error but a warning that this may be inefficient, and that `null_values` can be set to False for a speed-up. What do you think?

Comment (zazass8) [reply mlxtend/frequent_patterns/fpcommon.py:168]: Fixed! Please look at the latest commit.

File: mlxtend/frequent_patterns/tests/test_association_rules.py (hunk 5 above)
Comment (rasbt) [reply mlxtend/frequent_patterns/tests/test_association_rules.py]: Awesome, thanks for adding this. I'd say don't worry about the formatting. It's one of the caveats when using black. It's just in the tests so it's ok.

File: mlxtend/frequent_patterns/fpcommon.py [hunk 6]
```diff
@@ -104,14 +148,36 @@ def valid_input_check(df):
...
             "DataFrames with non-bool types result in worse computational"
             "performance and their support might be discontinued in the future."
             "Please use a DataFrame with bool type",
             DeprecationWarning,
         )
+
+        # If null_values is True but no NaNs are found, raise an error
+        has_nans = pd.isna(df).any().any()
+        if null_values and not has_nans:
+            warnings.warn(
```
Comment (rasbt) [inline mlxtend/frequent_patterns/fpcommon.py:171]: Looks great, thanks for the update!

**Review by rasbt — APPROVED** [review]
Awesome work! And thanks so much for addressing all the minor concerns I have.

**Comment by zazass8** [comment]
> Awesome work! And thanks so much for addressing all the minor concerns I have.

No worries at all, my pleasure in contributing. Until next time!

**Comment by d-kleine** [comment]
fixed #1004

**Comment by jraymartinez** [comment]
Memory usage regression after PR #1106

We noticed significant memory usage increase in our production pipeline using fpgrowth algorithm after recent updates. 

Details:
- Pipeline was running successfully weekly until November 8, 2024
- Same input data size and parameters throughout (no changes in data volume or fpgrowth parameters)
- Originally running on ml.c5n.9xlarge (96GB RAM)
- After failure, upgraded to ml.c5n.18xlarge (192GB RAM) but still getting memory errors
- Error message: "Please use an instance type with more memory, or reduce the size of job data processed on an instance"
- When pinned to mlxtend==0.23.1 (pre-PR version), pipeline runs successfully again on original instance size

This suggests potential memory regression after PR #1106 (merged Oct 23, 2024). Could you please investigate if there are any unintended memory usage increases from these changes?

Environment:
- AWS SageMaker
- Python 3.10
- Input data shape: (1832691, 11161)
- FPGrowth parameters: min_support=0.0005, max_len=15

**Comment by rasbt** [comment]
Thanks for letting us know. I will take a look at this in the next few weeks and see what's causing this. In the meantime, I recommend sticking to `mlxtend==0.23.1` as you suggested.

If you have any ideas @zazass8 let me know

**Comment by zazass8** [comment]
I investigated the issue. It seems that in `fpcommon.py` where the `disabled` array is initialised, multiple temporary arrays are created to generate the desired one. For that I already worked on an alternative solution which it could work.

Furthermore, on line 40 where I used `np.sum` and `np.logical_or` would consume a lot of memory for the same as above. For this, when I was working on the PR I actually used directly `np.nansum()` as I did in the next line. Where, theoretically it can solve the memory allocation issue as it directly computes what is desired. Although, several unit tests when the code was tested on CI were failing without me figuring out the source of the issue. 

Finally, in `generate_itemsets` maybe the fact that multiple if statements are implemented may cause again memory allocation. I couldn't think of any other way to simplify those lines in order to obtain the desired output.

@rasbt if you have any ideas, let me know as well. Thank you for pointing this out @jraymartinez.

**Comment by rasbt** [comment]
Thanks for the analysis @zazass8! I would say we should maybe investigate why `np.nansum` doesn't/didn't work. Maybe that is the main culprit here.
I currently only have access to a small laptop until December but can help try some things in December when I am back.

**Comment by billyes** [comment]
Why do we add num_itemsets as required input for association_rules? It can be inferred from df size so at least we should make it optional IMO.

**Comment by zazass8** [comment]
> Why do we add num_itemsets as required input for association_rules? It can be inferred from df size so at least we should make it optional IMO.
The reason it was added is because if we use the algorithm when the input dataset includes missing values the metrics that are generated along with the association rules are updated differently (please have a look at the updated formulae on `metric_dict`). You are right, it shouldn't be required for the case if `null_values=False` but I would suggest to instead set it as default equal to 1 in order to remain consistent with the implementation of the formulae. Then if you do this, also remove line 288 (`num_itemsets = 1`) because I noticed it is not really needed if we set it as default.
